File: backend-nestjs/src/property-operations/property-operations-public.service.ts
import { BadRequestException, GoneException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { createHash, randomUUID } from 'crypto';
import { In, Repository } from 'typeorm';
import {
  PropertyOperationsPublicAccess,
  PropertyOperationsRecord,
  PropertyOperationsSubmission,
} from './property-operations.entity';
import { PropertyOperationsIntegrationService } from './property-operations-integration.service';
import { PropertyOperationsService } from './property-operations.service';

@Injectable()
export class PropertyOperationsPublicService {
  constructor(
    @InjectRepository(PropertyOperationsPublicAccess)
    private readonly accessRepository: Repository<PropertyOperationsPublicAccess>,
    @InjectRepository(PropertyOperationsSubmission)
    private readonly submissionRepository: Repository<PropertyOperationsSubmission>,
    @InjectRepository(PropertyOperationsRecord)
    private readonly recordRepository: Repository<PropertyOperationsRecord>,
    private readonly operationsService: PropertyOperationsService,
    private readonly integrations: PropertyOperationsIntegrationService,
  ) {}

  async list(propertyId?: number) {
    const workspace = propertyId ? await this.operationsService.workspaceByPropertyId(propertyId) : null;
    const accesses = await this.accessRepository.find({
      where: workspace ? { workspaceId: workspace.id } : {},
      order: { createdAt: 'DESC' },
    });
    const workspaces = await Promise.all(
      [...new Set(accesses.map((item) => item.workspaceId))].map((id) => this.operationsService.workspaceById(id)),
    );
    const submissions = accesses.length
      ? await this.submissionRepository.find({ where: { accessId: In(accesses.map((item) => item.id)) } })
      : [];
    return accesses.map((access) => this.accessDto(
      access,
      workspaces.find((workspaceItem) => workspaceItem.id === access.workspaceId),
      submissions.filter((submission) => submission.accessId === access.id).length,
    ));
  }

  async create(input: any) {
    const workspace = await this.operationsService.workspaceByPropertyId(Number(input?.propertyId));
    const settings = await this.operationsService.getSettings();
    const token = `${randomUUID()}${randomUUID()}`.replace(/-/g, '');
    const expiryHours = Math.max(1, Number(input?.expiryHours ?? settings.defaultExpiryHours));
    const expiresAt = input?.expiresAt ? new Date(input.expiresAt) : new Date(Date.now() + expiryHours * 3_600_000);
    if (!Number.isFinite(expiresAt.getTime()) || expiresAt.getTime() <= Date.now()) {
      throw new BadRequestException('Public request expiry must be in the future.');
    }
    const recordId = input?.recordId ? Number(input.recordId) : null;
    const linkedRecord = recordId
      ? await this.recordRepository.findOne({ where: { id: recordId, workspaceId: workspace.id } })
      : null;
    if (recordId && !linkedRecord) throw new NotFoundException('The linked Property Operations record was not found.');
    const linkedAmount = linkedRecord && ['billing', 'finance'].includes(linkedRecord.moduleKey)
      ? Number(linkedRecord.amount ?? 0)
      : 0;
    const paymentAmount = input?.paymentAmount == null ? linkedAmount : Number(input.paymentAmount);

    const access = await this.accessRepository.save(this.accessRepository.create({
      workspaceId: workspace.id,
      recordId: linkedRecord?.id ?? null,
      moduleKey: String(input?.moduleKey ?? linkedRecord?.moduleKey ?? 'public-portals'),
      tokenHash: this.hash(token),
      accessToken: token,
      title: String(input?.title ?? linkedRecord?.title ?? 'Property request'),
      instructions: String(input?.instructions ?? linkedRecord?.description ?? ''),
      recipientLabel: String(input?.recipientLabel ?? linkedRecord?.contactLabel ?? ''),
      recipientName: String(input?.recipientName ?? linkedRecord?.contactName ?? ''),
      recipientEmail: String(input?.recipientEmail ?? linkedRecord?.contactEmail ?? ''),
      recipientPhone: String(input?.recipientPhone ?? linkedRecord?.contactPhone ?? ''),
      status: 'Active',
      formSchema: Array.isArray(input?.formSchema) ? input.formSchema : [],
      expiresAt,
      maxUses: Math.max(1, Number(input?.maxUses ?? settings.defaultMaxUses)),
      useCount: 0,
      oneTime: input?.oneTime ?? settings.defaultOneTime,
      allowFileUploads: input?.allowFileUploads ?? settings.allowFileUploads,
      paymentAmount: paymentAmount > 0 ? String(paymentAmount) : null,
      paymentCurrency: String(input?.paymentCurrency ?? settings.currency ?? 'USD').toUpperCase(),
      paymentVerified: linkedRecord?.status === 'Paid',
      paymentSessionId: null,
      paymentTokenHash: null,
      lastAccessedAt: null,
      completedAt: null,
    }));
    await this.operationsService.logActivity(
      'External access created',
      `${access.title} link created.`,
      workspace.id,
      linkedRecord?.id,
      { action: 'external-access-created', sourceModule: access.moduleKey, accessId: access.id },
    );
    return this.accessDto(access, workspace, 0);
  }

  async revoke(id: number) {
    const access = await this.accessRepository.findOne({ where: { id } });
    if (!access) throw new NotFoundException('Public request link was not found.');
    access.status = 'Revoked';
    access.completedAt = new Date();
    await this.accessRepository.save(access);
    await this.operationsService.logActivity(
      'External access revoked',
      `${access.title} link revoked.`,
      access.workspaceId,
      access.recordId ?? undefined,
      { action: 'external-access-revoked', sourceModule: access.moduleKey, accessId: access.id },
    );
    return { message: 'Public request link revoked.' };
  }

  async submissions(accessId: number) {
    const access = await this.accessRepository.findOne({ where: { id: accessId } });
    if (!access) throw new NotFoundException('Public request link was not found.');
    const items = await this.submissionRepository.find({ where: { accessId }, order: { createdAt: 'DESC' } });
    return items.map((item) => ({
      id: String(item.id),
      responderName: item.responderName,
      responderEmail: item.responderEmail,
      responderPhone: item.responderPhone,
      notes: item.notes,
      response: item.response,
      attachmentUrls: item.attachmentUrls,
      submittedAt: item.createdAt.toISOString(),
      createdAt: item.createdAt.toISOString(),
    }));
  }

  async getRequest(token: string) {
    const access = await this.requireReadableAccess(token);
    access.lastAccessedAt = new Date();
    await this.accessRepository.save(access);
    const workspace = await this.operationsService.workspaceById(access.workspaceId);
    const settings = await this.operationsService.getSettings();
    const snapshot: any = workspace.propertySnapshot ?? {};
    const record = access.recordId ? await this.recordRepository.findOne({ where: { id: access.recordId } }) : null;
    const paymentConfigured = await this.integrations.hasPaymentProvider();
    return {
      businessName: settings.businessName,
      logoUrl: settings.logoUrl,
      brandColor: settings.brandColor,
      currency: settings.currency,
      welcomeMessage: settings.welcomeMessage,
      termsText: settings.termsText,
      requireName: settings.requireName,
      requireEmail: settings.requireEmail,
      requirePhone: settings.requirePhone,
      allowFileUploads: access.allowFileUploads,
      showPropertyAddress: settings.showPropertyAddress,
      propertyTitle: String(snapshot.title ?? `Property ${workspace.propertyId}`),
      propertyLocation: String(snapshot.location ?? ''),
      moduleKey: access.moduleKey,
      title: access.title,
      instructions: access.instructions,
      recipientLabel: access.recipientLabel,
      formSchemaJson: JSON.stringify(access.formSchema ?? []),
      expiresAt: access.expiresAt.toISOString(),
      remainingUses: Math.max(0, access.maxUses - access.useCount),
      requestStatus: access.status,
      linkedRecordId: record ? String(record.id) : null,
      paymentEnabled: paymentConfigured && Boolean(access.paymentAmount && Number(access.paymentAmount) > 0),
      paymentAmount: access.paymentAmount == null ? null : Number(access.paymentAmount),
      paymentCurrency: access.paymentCurrency || settings.currency,
      paymentVerified: access.paymentVerified || record?.status === 'Paid',
      linkedRecordStatus: record?.status ?? null,
    };
  }

  async submit(token: string, input: any) {
    const access = await this.requireActiveAccess(token);
    const settings = await this.operationsService.getSettings();
    if (settings.requireName && !String(input?.responderName ?? '').trim()) throw new BadRequestException('Your name is required.');
    if (settings.requireEmail && !String(input?.responderEmail ?? '').trim()) throw new BadRequestException('Your email is required.');
    if (settings.requirePhone && !String(input?.responderPhone ?? '').trim()) throw new BadRequestException('Your phone number is required.');
    const responsePayload = input?.response && typeof input.response === 'object' ? input.response : {};
    const attachmentUrls = Array.isArray(input?.attachmentUrls) ? input.attachmentUrls.map(String) : [];
    this.validateRequiredFields(access.formSchema, responsePayload, attachmentUrls);

    const submission = await this.submissionRepository.save(this.submissionRepository.create({
      accessId: access.id,
      responderName: String(input?.responderName ?? ''),
      responderEmail: String(input?.responderEmail ?? ''),
      responderPhone: String(input?.responderPhone ?? ''),
      notes: String(input?.notes ?? ''),
      response: responsePayload,
      attachmentUrls,
    }));
    access.useCount += 1;
    access.lastAccessedAt = new Date();
    if (access.oneTime || access.useCount >= access.maxUses) {
      access.status = 'Completed';
      access.completedAt = new Date();
    }
    await this.accessRepository.save(access);

    if (access.recordId) {
      const record = await this.recordRepository.findOne({ where: { id: access.recordId } });
      if (record) {
        record.status = settings.autoCloseRecordOnSubmit ? 'Completed' : 'Submitted';
        record.completedAt = settings.autoCloseRecordOnSubmit ? new Date() : null;
        record.payload = {
          ...(record.payload ?? {}),
          latestSubmissionId: submission.id,
          latestSubmissionAt: submission.createdAt.toISOString(),
          latestSubmission: responsePayload,
        };
        await this.recordRepository.save(record);
      }
    }
    await this.operationsService.logActivity(
      'External response submitted',
      `Response submitted for ${access.title}.`,
      access.workspaceId,
      access.recordId ?? undefined,
      { action: 'external-response-submitted', sourceModule: access.moduleKey, accessId: access.id, submissionId: submission.id },
    );
    return { id: String(submission.id), status: access.status, submittedAt: submission.createdAt.toISOString() };
  }

  async getStatus(token: string) {
    const access = await this.requireReadableAccess(token);
    if (!access.recordId) throw new NotFoundException('This public request is not linked to a record.');
    const record = await this.recordRepository.findOne({ where: { id: access.recordId } });
    if (!record) throw new NotFoundException('The linked Property Operations record was not found.');
    const history = await this.submissionRepository.find({ where: { accessId: access.id }, order: { createdAt: 'ASC' } });
    return {
      id: String(record.id), moduleKey: record.moduleKey, recordType: record.recordType,
      title: record.title, description: record.description, status: record.status, priority: record.priority,
      assignedTo: record.assignedTo, dueAt: record.dueAt?.toISOString() ?? null,
      amount: record.amount == null ? null : Number(record.amount), attachments: record.attachments ?? [],
      history: history.map((item) => ({
        id: String(item.id), responderName: item.responderName, notes: item.notes,
        response: item.response, attachmentUrls: item.attachmentUrls, createdAt: item.createdAt.toISOString(),
      })),
    };
  }

  async updateStatus(token: string, input: any) {
    const access = await this.requireActiveAccess(token);
    if (!access.recordId) throw new NotFoundException('This public request is not linked to a record.');
    const record = await this.recordRepository.findOne({ where: { id: access.recordId } });
    if (!record) throw new NotFoundException('The linked Property Operations record was not found.');
    const allowed = ['Acknowledged', 'In progress', 'Waiting', 'Resolved', 'Completed'];
    if (input?.status && !allowed.includes(String(input.status))) throw new BadRequestException('Unsupported public status.');
    if (input?.status) record.status = String(input.status);
    if (input?.notes) record.payload = { ...(record.payload ?? {}), publicStatusNote: String(input.notes), publicStatusUpdatedAt: new Date().toISOString() };
    record.completedAt = ['Resolved', 'Completed'].includes(record.status) ? new Date() : null;
    await this.recordRepository.save(record);
    await this.operationsService.logActivity(
      'Public status updated',
      `${record.title} marked ${record.status}.`,
      access.workspaceId,
      record.id,
      { action: 'public-status-updated', sourceModule: record.moduleKey, status: record.status },
    );
    return this.getStatus(token);
  }

  async createCheckout(token: string, successUrl: string, cancelUrl: string) {
    const access = await this.requireActiveAccess(token);
    if (!access.recordId) throw new NotFoundException('No billing record is linked to this request.');
    const record = await this.recordRepository.findOne({ where: { id: access.recordId } });
    if (!record || !['billing', 'finance'].includes(record.moduleKey)) throw new BadRequestException('Only billing or finance records can be paid.');
    const amount = Number(access.paymentAmount ?? record.amount ?? 0);
    if (amount <= 0) throw new BadRequestException('A positive billing amount is required.');
    const checkout = await this.integrations.createCheckout({
      amount,
      currency: access.paymentCurrency || 'USD',
      title: record.title,
      reference: String(record.id),
      successUrl,
      cancelUrl,
    });
    access.paymentSessionId = checkout.sessionId;
    access.paymentTokenHash = this.hash(checkout.paymentToken);
    await this.accessRepository.save(access);
    await this.operationsService.logActivity(
      'Checkout created',
      `Checkout created for ${record.title}.`,
      access.workspaceId,
      record.id,
      { action: 'checkout-created', sourceModule: record.moduleKey, amount, currency: access.paymentCurrency },
    );
    return { checkoutUrl: checkout.checkoutUrl, sessionId: checkout.sessionId, paymentToken: checkout.paymentToken };
  }

  async verifyCheckout(token: string, sessionId: string, paymentToken: string) {
    const access = await this.requireReadableAccess(token);
    if (!access.recordId) throw new NotFoundException('No billing record is linked to this request.');
    if (access.paymentSessionId && sessionId && access.paymentSessionId !== sessionId) throw new BadRequestException('Payment session does not match this request.');
    if (access.paymentTokenHash && paymentToken && access.paymentTokenHash !== this.hash(paymentToken)) throw new BadRequestException('Payment token is invalid.');
    const reference = sessionId || access.paymentSessionId || String(access.recordId);
    const verification = await this.integrations.verifyCheckout(reference, paymentToken);
    const record = await this.recordRepository.findOne({ where: { id: access.recordId } });
    if (!record) throw new NotFoundException('The linked billing record was not found.');
    if (verification.paid) {
      access.paymentVerified = true;
      access.status = 'Completed';
      access.completedAt = new Date();
      record.status = 'Paid';
      record.completedAt = new Date();
      record.payload = { ...(record.payload ?? {}), paymentReference: reference, paymentStatus: 'Paid' };
      await Promise.all([this.accessRepository.save(access), this.recordRepository.save(record)]);
      await this.operationsService.logActivity(
        'Payment verified',
        `${record.title} marked paid.`,
        access.workspaceId,
        record.id,
        { action: 'payment-verified', sourceModule: record.moduleKey, reference },
      );
    }
    return verification;
  }

  async assertUploadAllowed(token: string) {
    const access = await this.requireActiveAccess(token);
    if (!access.allowFileUploads) throw new BadRequestException('File uploads are disabled for this request.');
    return access;
  }

  private async requireActiveAccess(token: string) {
    const access = await this.requireReadableAccess(token);
    if (access.status !== 'Active') throw new GoneException('This public request link is no longer active.');
    if (access.useCount >= access.maxUses) throw new GoneException('This public request link has reached its usage limit.');
    return access;
  }

  private async requireReadableAccess(token: string) {
    const access = await this.requireAccess(token);
    if (access.status === 'Revoked') throw new GoneException('This public request link was revoked.');
    if (access.status === 'Expired' || access.expiresAt.getTime() <= Date.now()) {
      if (access.status !== 'Expired') {
        access.status = 'Expired';
        await this.accessRepository.save(access);
      }
      throw new GoneException('This public request link has expired.');
    }
    return access;
  }

  private async requireAccess(token: string) {
    const cleanToken = String(token ?? '').trim();
    if (!cleanToken) throw new NotFoundException('Public request link was not found.');
    const access = await this.accessRepository.findOne({ where: { tokenHash: this.hash(cleanToken) } });
    if (!access) throw new NotFoundException('Public request link was not found.');
    return access;
  }

  private accessDto(access: PropertyOperationsPublicAccess, workspace: any, submissionCount: number) {
    const snapshot: any = workspace?.propertySnapshot ?? {};
    const publicBaseUrl = String(process.env.FRONTEND_URL ?? process.env.APP_URL ?? 'http://localhost:3000').replace(/\/$/, '');
    const publicUrl = `${publicBaseUrl}/property-request/${access.accessToken}`;
    return {
      id: String(access.id), propertyId: workspace?.propertyId ?? 0,
      recordId: access.recordId == null ? null : String(access.recordId), moduleKey: access.moduleKey,
      title: access.title, instructions: access.instructions, recipientLabel: access.recipientLabel,
      recipientName: access.recipientName, recipientEmail: access.recipientEmail, recipientPhone: access.recipientPhone,
      status: access.status, formSchemaJson: JSON.stringify(access.formSchema ?? []),
      expiresAt: access.expiresAt.toISOString(), maxUses: access.maxUses, useCount: access.useCount,
      oneTime: access.oneTime, allowFileUploads: access.allowFileUploads,
      createdAt: access.createdAt.toISOString(), lastAccessedAt: access.lastAccessedAt?.toISOString() ?? null,
      completedAt: access.completedAt?.toISOString() ?? null,
      propertyTitle: String(snapshot.title ?? `Property ${workspace?.propertyId ?? ''}`),
      propertyLocation: String(snapshot.location ?? ''),
      publicUrl,
      accessToken: access.accessToken,
      qrDataUrl: `https://api.qrserver.com/v1/create-qr-code/?size=360x360&margin=2&data=${encodeURIComponent(publicUrl)}`,
      submissionCount,
      paymentAmount: access.paymentAmount == null ? null : Number(access.paymentAmount),
      paymentCurrency: access.paymentCurrency,
      paymentVerified: access.paymentVerified,
    };
  }

  private validateRequiredFields(schema: unknown[], response: Record<string, unknown>, attachments: string[]) {
    for (const raw of Array.isArray(schema) ? schema : []) {
      const field = raw && typeof raw === 'object' ? raw as Record<string, unknown> : {};
      if (field.required !== true) continue;
      const key = String(field.key ?? '');
      const type = String(field.type ?? 'text');
      const value = response[key];
      const missing = type === 'file'
        ? !attachments.length && !value
        : type === 'checkbox'
          ? value !== true
          : value == null || String(value).trim() === '';
      if (missing) throw new BadRequestException(`${String(field.label ?? key || 'Required field')} is required.`);
    }
  }

  private hash(value: string) {
    return createHash('sha256').update(value).digest('hex');
  }
}

File: backend/Services/PropertyOperationsPublicService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Data;
using Entities;
using Microsoft.EntityFrameworkCore;
using Models;

namespace Services;

public sealed class PropertyOperationsPublicService(AppDbContext db)
{
    public async Task<PropertyOperationsSettingsRecord> GetSettingsAsync(CancellationToken ct = default)
    {
        var settings = await db.PropertyOperationsSettings.FirstOrDefaultAsync(item => item.Id == 1, ct);
        if (settings is not null) return settings;

        settings = new PropertyOperationsSettingsRecord { Id = 1 };
        db.PropertyOperationsSettings.Add(settings);
        await db.SaveChangesAsync(ct);
        return settings;
    }

    public async Task<PropertyOperationsSettingsRecord> UpdateSettingsAsync(UpdatePropertyOperationsSettingsRequest req, CancellationToken ct = default)
    {
        var settings = await GetSettingsAsync(ct);
        settings.BusinessName = Text(req.BusinessName, "Property Operations", 200);
        settings.LogoUrl = Text(req.LogoUrl, string.Empty, 1000);
        settings.BrandColor = Text(req.BrandColor, "#111827", 20);
        settings.PublicBaseUrl = Text(req.PublicBaseUrl, string.Empty, 1000).TrimEnd('/');
        settings.DefaultExpiryHours = Math.Clamp(req.DefaultExpiryHours, 1, 8760);
        settings.DefaultMaxUses = Math.Clamp(req.DefaultMaxUses, 1, 1000);
        settings.DefaultOneTime = req.DefaultOneTime;
        settings.RequireName = req.RequireName;
        settings.RequireEmail = req.RequireEmail;
        settings.RequirePhone = req.RequirePhone;
        settings.AllowFileUploads = req.AllowFileUploads;
        settings.ShowPropertyAddress = req.ShowPropertyAddress;
        settings.AutoCloseRecordOnSubmit = req.AutoCloseRecordOnSubmit;
        settings.NotifyAdminOnSubmit = req.NotifyAdminOnSubmit;
        settings.WelcomeMessage = Text(req.WelcomeMessage, string.Empty, 5000);
        settings.TermsText = Text(req.TermsText, string.Empty, 10000);
        settings.EmailSubjectTemplate = Text(req.EmailSubjectTemplate, string.Empty, 1000);
        settings.EmailBodyTemplate = Text(req.EmailBodyTemplate, string.Empty, 10000);
        settings.SmsTemplate = Text(req.SmsTemplate, string.Empty, 2000);
        settings.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(ct);
        await AddActivityAsync(null, "organization", "settings-updated", "Settings", 1, "Admin", "Property Operations settings updated.", "{}", ct);
        return settings;
    }

    public async Task<(PropertyOperationsPublicAccess Item, string Token, string Url)> CreateAccessAsync(CreatePropertyOperationsPublicAccessRequest req, CancellationToken ct = default)
    {
        var workspace = await db.PropertyOperationsWorkspaces
            .Include(item => item.Property)
            .FirstOrDefaultAsync(item => item.PropertyId == req.PropertyId, ct)
            ?? throw new ArgumentException("Import the property before creating a public request.");

        var settings = await GetSettingsAsync(ct);
        var token = Convert.ToHexString(RandomNumberGenerator.GetBytes(32)).ToLowerInvariant();
        var item = new PropertyOperationsPublicAccess
        {
            WorkspaceId = workspace.Id,
            RecordId = req.RecordId,
            ModuleKey = Text(req.ModuleKey, "portfolio", 80).ToLowerInvariant(),
            Title = Text(req.Title, "Property request", 240),
            Instructions = Text(req.Instructions, string.Empty, 10000),
            RecipientLabel = Text(req.RecipientLabel, "External recipient", 120),
            RecipientName = Text(req.RecipientName, string.Empty, 160),
            RecipientEmail = Text(req.RecipientEmail, string.Empty, 240).ToLowerInvariant(),
            RecipientPhone = Text(req.RecipientPhone, string.Empty, 80),
            TokenHash = HashToken(token),
            FormSchemaJson = Json(req.FormSchemaJson, "[]"),
            ExpiresAt = DateTime.UtcNow.AddHours(Math.Clamp(req.ExpiryHours ?? settings.DefaultExpiryHours, 1, 8760)),
            MaxUses = Math.Clamp(req.MaxUses ?? settings.DefaultMaxUses, 1, 1000),
            OneTime = req.OneTime ?? settings.DefaultOneTime,
            AllowFileUploads = req.AllowFileUploads ?? settings.AllowFileUploads,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };
        db.PropertyOperationsPublicAccessLinks.Add(item);
        await db.SaveChangesAsync(ct);
        var baseUrl = string.IsNullOrWhiteSpace(settings.PublicBaseUrl) ? string.Empty : settings.PublicBaseUrl;
        var url = $"{baseUrl}/property-request/{token}";
        await AddActivityAsync(workspace.Id, item.ModuleKey, "public-link-created", "PublicAccess", item.Id, "Admin", $"Public request created for {item.RecipientLabel}.", "{}", ct);
        return (item, token, url);
    }

    public async Task<List<PropertyOperationsPublicAccess>> GetAccessLinksAsync(int? propertyId, CancellationToken ct = default)
    {
        var query = db.PropertyOperationsPublicAccessLinks
            .AsNoTracking()
            .Include(item => item.Workspace).ThenInclude(item => item.Property)
            .Include(item => item.Submissions)
            .AsQueryable();
        if (propertyId.HasValue) query = query.Where(item => item.Workspace.PropertyId == propertyId.Value);
        return await query.OrderByDescending(item => item.CreatedAt).ToListAsync(ct);
    }

    public async Task<List<PropertyOperationsPublicSubmission>> GetSubmissionsAsync(int accessId, CancellationToken ct = default) =>
        await db.PropertyOperationsPublicSubmissions.AsNoTracking().Where(item => item.PublicAccessId == accessId).OrderByDescending(item => item.SubmittedAt).ToListAsync(ct);

    public async Task RevokeAsync(int id, CancellationToken ct = default)
    {
        var item = await db.PropertyOperationsPublicAccessLinks.FirstOrDefaultAsync(link => link.Id == id, ct);
        if (item is null) return;
        item.Status = "Revoked";
        item.RevokedAt = DateTime.UtcNow;
        item.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(ct);
        await AddActivityAsync(item.WorkspaceId, item.ModuleKey, "public-link-revoked", "PublicAccess", item.Id, "Admin", "Public request link revoked.", "{}", ct);
    }

    public async Task<(PropertyOperationsPublicAccess Access, PropertyOperationsSettingsRecord Settings)> GetPublicAsync(string token, CancellationToken ct = default)
    {
        var access = await FindActiveAsync(token, false, ct);
        access.LastAccessedAt = DateTime.UtcNow;
        access.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(ct);
        return (access, await GetSettingsAsync(ct));
    }

    public async Task<PropertyOperationsPublicSubmission> SubmitAsync(string token, SubmitPropertyOperationsPublicRequest req, CancellationToken ct = default)
    {
        var access = await FindActiveAsync(token, true, ct);
        var settings = await GetSettingsAsync(ct);
        if (settings.RequireName && string.IsNullOrWhiteSpace(req.ResponderName)) throw new ArgumentException("Name is required.");
        if (settings.RequireEmail && string.IsNullOrWhiteSpace(req.ResponderEmail)) throw new ArgumentException("Email is required.");
        if (settings.RequirePhone && string.IsNullOrWhiteSpace(req.ResponderPhone)) throw new ArgumentException("Phone is required.");

        var submission = new PropertyOperationsPublicSubmission
        {
            PublicAccessId = access.Id,
            ResponderName = Text(req.ResponderName, string.Empty, 160),
            ResponderEmail = Text(req.ResponderEmail, string.Empty, 240).ToLowerInvariant(),
            ResponderPhone = Text(req.ResponderPhone, string.Empty, 80),
            Notes = Text(req.Notes, string.Empty, 10000),
            ResponseJson = Json(req.ResponseJson, "{}"),
            AttachmentUrlsJson = Json(req.AttachmentUrlsJson, "[]"),
            SubmittedAt = DateTime.UtcNow,
        };
        db.PropertyOperationsPublicSubmissions.Add(submission);
        access.UseCount += 1;
        access.LastAccessedAt = DateTime.UtcNow;
        access.UpdatedAt = DateTime.UtcNow;
        if (access.OneTime || access.UseCount >= access.MaxUses)
        {
            access.Status = "Completed";
            access.CompletedAt = DateTime.UtcNow;
        }
        if (settings.AutoCloseRecordOnSubmit && access.RecordId.HasValue)
        {
            var record = await db.PropertyOperationsRecords.FirstOrDefaultAsync(item => item.Id == access.RecordId.Value, ct);
            if (record is not null)
            {
                record.Status = "Completed";
                record.UpdatedAt = DateTime.UtcNow;
            }
        }
        await db.SaveChangesAsync(ct);
        await AddActivityAsync(access.WorkspaceId, access.ModuleKey, "anonymous-submission", "PublicSubmission", submission.Id, "External", $"Anonymous response submitted for {access.Title}.", submission.ResponseJson, ct);
        return submission;
    }

    public async Task<PropertyOperationsAnalyticsResponse> GetAnalyticsAsync(CancellationToken ct = default)
    {
        var now = DateTime.UtcNow;
        var records = db.PropertyOperationsRecords.AsNoTracking();
        var response = new PropertyOperationsAnalyticsResponse
        {
            ImportedProperties = await db.PropertyOperationsWorkspaces.CountAsync(ct),
            TotalRecords = await records.CountAsync(ct),
            OpenRecords = await records.CountAsync(item => item.Status != "Completed" && item.Status != "Closed" && item.Status != "Paid", ct),
            OverdueRecords = await records.CountAsync(item => item.DueAt < now && item.Status != "Completed" && item.Status != "Closed" && item.Status != "Paid", ct),
            CompletedRecords = await records.CountAsync(item => item.Status == "Completed" || item.Status == "Closed" || item.Status == "Paid", ct),
            ActivePublicLinks = await db.PropertyOperationsPublicAccessLinks.CountAsync(item => item.Status == "Active" && item.ExpiresAt > now, ct),
            CompletedPublicLinks = await db.PropertyOperationsPublicAccessLinks.CountAsync(item => item.Status == "Completed", ct),
            AnonymousSubmissions = await db.PropertyOperationsPublicSubmissions.CountAsync(ct),
            TotalIncome = await records.Where(item => item.ModuleKey == "finance" && item.RecordType.ToLower().Contains("income")).SumAsync(item => item.Amount ?? 0, ct),
            TotalExpense = await records.Where(item => item.ModuleKey == "finance" && item.RecordType.ToLower().Contains("expense")).SumAsync(item => item.Amount ?? 0, ct),
            RecordsByModule = await records.GroupBy(item => item.ModuleKey).ToDictionaryAsync(group => group.Key, group => group.Count(), ct),
            RecordsByStatus = await records.GroupBy(item => item.Status).ToDictionaryAsync(group => group.Key, group => group.Count(), ct),
        };
        return response;
    }

    public async Task<List<PropertyOperationsActivityResponse>> GetActivitiesAsync(int? propertyId, CancellationToken ct = default)
    {
        var query = db.PropertyOperationsActivities.AsNoTracking().AsQueryable();
        if (propertyId.HasValue)
        {
            var workspaceId = await db.PropertyOperationsWorkspaces.Where(item => item.PropertyId == propertyId.Value).Select(item => (int?)item.Id).FirstOrDefaultAsync(ct);
            query = query.Where(item => item.WorkspaceId == workspaceId);
        }
        return await query.OrderByDescending(item => item.CreatedAt).Take(250).Select(item => new PropertyOperationsActivityResponse
        {
            Id = item.Id,
            ModuleKey = item.ModuleKey,
            Action = item.Action,
            EntityType = item.EntityType,
            EntityId = item.EntityId,
            ActorType = item.ActorType,
            Summary = item.Summary,
            MetadataJson = item.MetadataJson,
            CreatedAt = item.CreatedAt,
        }).ToListAsync(ct);
    }

    private async Task<PropertyOperationsPublicAccess> FindActiveAsync(string token, bool forSubmission, CancellationToken ct)
    {
        var hash = HashToken(token);
        var item = await db.PropertyOperationsPublicAccessLinks
            .Include(access => access.Workspace).ThenInclude(workspace => workspace.Property)
            .FirstOrDefaultAsync(access => access.TokenHash == hash, ct)
            ?? throw new KeyNotFoundException("This request link is invalid.");
        if (item.Status == "Revoked") throw new InvalidOperationException("This request link was revoked.");
        if (item.ExpiresAt <= DateTime.UtcNow)
        {
            item.Status = "Expired";
            item.UpdatedAt = DateTime.UtcNow;
            await db.SaveChangesAsync(ct);
            throw new InvalidOperationException("This request link has expired.");
        }
        if (item.Status == "Completed" || item.UseCount >= item.MaxUses) throw new InvalidOperationException("This request has already been completed.");
        if (forSubmission && item.OneTime && item.UseCount > 0) throw new InvalidOperationException("This one-time request has already been used.");
        return item;
    }

    private async Task AddActivityAsync(int? workspaceId, string moduleKey, string action, string entityType, int? entityId, string actorType, string summary, string metadataJson, CancellationToken ct)
    {
        db.PropertyOperationsActivities.Add(new PropertyOperationsActivity
        {
            WorkspaceId = workspaceId,
            ModuleKey = moduleKey,
            Action = action,
            EntityType = entityType,
            EntityId = entityId,
            ActorType = actorType,
            Summary = summary,
            MetadataJson = Json(metadataJson, "{}"),
            CreatedAt = DateTime.UtcNow,
        });
        await db.SaveChangesAsync(ct);
    }

    public static PropertyOperationsSettingsResponse Map(PropertyOperationsSettingsRecord item) => new()
    {
        BusinessName = item.BusinessName,
        LogoUrl = item.LogoUrl,
        BrandColor = item.BrandColor,
        PublicBaseUrl = item.PublicBaseUrl,
        DefaultExpiryHours = item.DefaultExpiryHours,
        DefaultMaxUses = item.DefaultMaxUses,
        DefaultOneTime = item.DefaultOneTime,
        RequireName = item.RequireName,
        RequireEmail = item.RequireEmail,
        RequirePhone = item.RequirePhone,
        AllowFileUploads = item.AllowFileUploads,
        ShowPropertyAddress = item.ShowPropertyAddress,
        AutoCloseRecordOnSubmit = item.AutoCloseRecordOnSubmit,
        NotifyAdminOnSubmit = item.NotifyAdminOnSubmit,
        WelcomeMessage = item.WelcomeMessage,
        TermsText = item.TermsText,
        EmailSubjectTemplate = item.EmailSubjectTemplate,
        EmailBodyTemplate = item.EmailBodyTemplate,
        SmsTemplate = item.SmsTemplate,
        UpdatedAt = item.UpdatedAt,
    };

    public static PropertyOperationsPublicAccessResponse Map(PropertyOperationsPublicAccess item, string? token = null, string? url = null) => new()
    {
        Id = item.Id,
        PropertyId = item.Workspace.PropertyId,
        RecordId = item.RecordId,
        ModuleKey = item.ModuleKey,
        Title = item.Title,
        Instructions = item.Instructions,
        RecipientLabel = item.RecipientLabel,
        RecipientName = item.RecipientName,
        RecipientEmail = item.RecipientEmail,
        RecipientPhone = item.RecipientPhone,
        Status = item.Status,
        FormSchemaJson = item.FormSchemaJson,
        ExpiresAt = item.ExpiresAt,
        MaxUses = item.MaxUses,
        UseCount = item.UseCount,
        OneTime = item.OneTime,
        AllowFileUploads = item.AllowFileUploads,
        CreatedAt = item.CreatedAt,
        LastAccessedAt = item.LastAccessedAt,
        CompletedAt = item.CompletedAt,
        PropertyTitle = item.Workspace.Property.Title,
        PropertyLocation = item.Workspace.Property.ExactLocation ?? item.Workspace.Property.Location ?? string.Empty,
        PublicUrl = url,
        AccessToken = token,
        SubmissionCount = item.Submissions.Count,
    };

    private static string HashToken(string token) => Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(token))).ToLowerInvariant();
    private static string Text(string? value, string fallback, int max) { var text = string.IsNullOrWhiteSpace(value) ? fallback : value.Trim(); return text.Length <= max ? text : text[..max]; }
    private static string Json(string? value, string fallback) { var text = string.IsNullOrWhiteSpace(value) ? fallback : value; JsonDocument.Parse(text).Dispose(); return text; }
}
